transformer_engine/pytorch/attention.py

```diff
@@ -714,8 +917,21 @@ class DotProductAttention(torch.nn.Module):
                 number of key-value channels.
     attention_dropout: float, default = 0.0
                       dropout probability for the dropout op during multi-head attention.
-    attn_mask_type: {'causal', 'padding'}, default = `causal`
-                   type of attention mask passed into softmax operation.
+    attn_mask_type: str, default = `causal`
+                   type of attention mask passed into softmax operation, options are "`causal`",
+                   "`padding`", "`arbitrary`", "`no_mask`". For the "`causal`" mask,
+                   TransformerEngine calculates and applies an upper triangular mask to
+                   the softmax input. An "`arbitrary`" mask is an arbitrary user defined mask
+                   broadcastable to the shape of softmax input. The "`padding`" mask is used
+                   for providing locations of padded tokens in the batch, which should be of
+                   the shape [batch_size, 1, 1, seq_len]. No mask is applied for the "`no_mask`"
+                   option. For the `"arbitrary"` and `"padding"` mask types, the argument
+                   :attr:`attention_mask` must be passed into `forward` call. The "`causal`"
+                   mask can also be applied in conjunction with "`padding`" mask by passing
+                   in multiple mask type as a comma separated string, for example,
+                   `attn_mask_type="causal,padding"`.
```

Comment:
Maybe in the future we can separate these mask types into two groups like PyTorch did: one for padding/arbitrary, which accepts a mask tensor input but the algorithm for padding or arbitrary would be different; and the other group is for is_causal, which doesn't need a mask tensor. We can just use two separate flags to represent them, on all levels of the implementation (DPA, Flash/FusedAttention, C level). 
https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html